### PR TITLE
Treat `*.bat` files as valid executables when building on Windows OS

### PR DESCRIPTION
**Pull request by andreiled**
1. Add new `withExtension` method version in the `FileUtils` accepting
   an additional vararg array of alternative acceptable extensions to
   verify the provided file path against before resorting to replacing
   original provided file's extension with the provided replacement
   extension
2. Update `OperatingSystem.Windows.getExecutableName` to pass `.bat`
   to the `FileUtils.withExtension` method call as an alternative
   acceptable extension in addition to `.exe`

Fixes gradle/gradle-native#1075 Gradle does not recognize bat files as executables on Windows preventing using drop-in replacements for gcc/clang

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/base-services/src/main/java/org/gradle/internal/FileUtils.java [hunk 1]
```diff
@@ -173,12 +173,29 @@ private static boolean endsWithIgnoreCase(String subject, String suffix) {
      * @return the transformed path
      */
     public static String withExtension(String filePath, String extension) {
-        if (filePath.toLowerCase().endsWith(extension)) {
+        return withExtension(filePath, extension, new String[0]);
+    }
+
+    public static String withExtension(String filePath, String extension, String... alternativeExts) {
+        if (endsWithOneOf(filePath.toLowerCase(), extension, alternativeExts)) {
```
Comment (andreiled) [inline subprojects/base-services/src/main/java/org/gradle/internal/FileUtils.java:180]: At first I tried to implement this as follows:
```java
if (filePath.toLowerCase().endsWith(extension)
        || Arrays.stream(alternativeExts).anyMatch((altExt) -> filePath.toLowerCase().endsWith(extension))) {
```
but that resulted in an error that lambda is no supported with source level 6, so just to double-check I tried adding following to [`base-services.gradle.kts`](https://github.com/gradle/gradle/blob/master/subprojects/base-services/base-services.gradle.kts):
```kt
java {
    sourceCompatibility = JavaVersion.VERSION_1_8
    targetCompatibility = JavaVersion.VERSION_1_8
}
```
which resulted in a bunch of errors like this one when building the _base-services_ project using JDK 11.0.7:
```
D:\GIT\gradle\subprojects\base-services\src\main\java\org\gradle\api\specs\CompositeSpec.java:40: warning: [unchecked] Possible heap pollution from parameterized vararg type Spec<? super T>
    protected CompositeSpec(Spec<? super T>... specs) {
                                               ^
  where T is a type-variable:
    T extends Object declared in class CompositeSpec
error: warnings found and -Werror specified
```

Could anyone confirm that _base-services_ subproject is indeed built with Java source level 6 by design and I am not missing anything when it comes to building Gradle itself?

**Comment by JLLeitschuh** [comment]
@bot-gradle test this please

**Comment by JLLeitschuh** [comment]
@andreiled Please rebase these changes on master

**Comment by JLLeitschuh** [comment]
@Mergifyio rebase

**Comment by JLLeitschuh** [comment]
@bot-gradle test this please

**Comment by andreiled** [comment]
@JLLeitschuh, looks like the build failed with this exception is the primary cause:
```
installation("build/install/main", OperatingSystem.forName(installOS)).assertInstalled()
|                                  |               |       |           |
|                                  |               |       windows     java.lang.AssertionError: /mnt/tcagent1/work/e67123fb5b9af0ac/subprojects/platform-native/build/tmp/test files/InstallExec.Test/can_create_._one/5z2pc/build/install/main/lib/main.bat is not a file
|                                  |               |                   	at org.gradle.test.fixtures.file.TestFile.assertIsFile(TestFile.java:429)
|                                  |               |                   	at org.gradle.nativeplatform.fixtures.NativeInstallationFixture.assertInstalled(NativeInstallationFixture.groovy:52)
|                                  |               |                   	at org.gradle.nativeplatform.platform.InstallExecutableIntegrationTest.can create installation for a different os than the current one(InstallExecutableIntegrationTest.groovy:92)
|                                  |               Linux 4.4.0-1052-aws amd64
|                                  class org.gradle.internal.os.OperatingSystem
<org.gradle.nativeplatform.fixtures.NativeInstallationFixture@5a50700c installDir=/mnt/tcagent1/work/e67123fb5b9af0ac/subprojects/platform-native/build/tmp/test files/InstallExec.Test/can_create_._one/5z2pc/build/install/main os=Linux 4.4.0-1052-aws amd64>
```

For example:
* [InstallExecutableIntegrationTest.can create installation for a different os than the current one](https://builds.gradle.org/viewLog.html?buildId=39573054&tab=buildResultsDiv&buildTypeId=Gradle_Check_Quick_1_bucket47)
* [CppLibraryPublishingIntegrationTest.can publish a library and its dependencies to a Maven repository when multiple target architectures are specified](https://builds.gradle.org/viewLog.html?buildId=39573096&tab=buildResultsDiv&buildTypeId=Gradle_Check_Quick_2_bucket38#testNameId-4484704368313223265)

Assuming this is caused by my change (unfortunately, I do not know the Gradle code good enough neither to guess how nor to argue against it), would you be able to point me to an instruction how I can run failed Integration Tests locally in order to investigate the issue?

**Comment by JLLeitschuh** [comment]
@blindpirate  thoughts?

**Comment by blindpirate** [comment]
@andreiled Looking at the failing test and your change, it's very likely to be caused by your change. To rerun the failing test locally, you can check out the failing task in the build log, in your case, 

```
Task :platform-native:embeddedIntegTest FAILED
```

So it's `./gradlew :platform-native:embeddedIntegTest --tests InstallExecutableIntegrationTest`, note this is failing on Linux, so you might get a successful build if you run it on Windows.

**Comment by andreiled** [comment]
Thanks for the info, @blindpirate.
If I don't manage to reproduce the issue on Windows, I will try running the test in a WSL container.

**Pull request closed without merging**